app/controllers/review_item/donor_messages.js

```diff
@@ -1,7 +1,195 @@
+        return [];
+      }
+      var itemId = parseInt(this.get("item.id"), 10);
+      return this.get("allVersions")
+        .filterBy("itemId", itemId)
+        .filterBy("itemType", "Item");
+    }
+  ),
+
+  packageVersions: Ember.computed(
+    "item.packages",
+    "allVersions.[]",
+    "isItemThread",
+    function() {
+      if (!this.get("isItemThread")) {
+        return [];
+      }
+      var packageIds = (this.get("item.packages") || []).mapBy("id");
+      return this.get("allVersions")
+        .filterBy("itemType", "Package")
+        .filter(function(log) {
+          return (
+            packageIds.indexOf(String(log.get("itemId"))) >= 0 &&
+            ["received", "missing"].indexOf(log.get("state")) >= 0
+          );
+        });
+    }
+  ),
+
+  allVersions: Ember.computed(function() {
+    return this.get("store").peekAll("version");
+  }),
+
+  offerVersions: Ember.computed(
+    "allVersions.[]",
+    "offer.id",
+    "isItemThread",
+    function() {
+      if (this.get("isItemThread")) {
+        return [];
+      }
+      var offerId = parseInt(this.get("offer.id"), 10);
+      return this.get("allVersions")
+        .filterBy("itemType", "Offer")
+        .filterBy("itemId", offerId);
+    }
+  ),
+
+  groupBy: function(content, key) {
+    var result = [];
+    var object, value;
+
+    content.forEach(function(item) {
+      value = item.get ? item.get(key) : item[key];
+      object = result.findBy("value", value);
+      if (!object) {
+        object = {
+          value: value,
+          items: []
+        };
+        result.push(object);
+      }
+      return object.items.push(item);
+    });
+    return result.getEach("items");
+  },
+
+  markConversationAsRead() {
+    this.get("messages")
+      .filterBy("state", "unread")
+      .forEach(m => this.get("messagesUtil").markRead(m));
+  },
+
+  actions: {
+    async cannedMessageLookup() {
+      try {
+        const text = await this.get("messageService").cannedMessageLookup();
+        this.set("body", text);
+        console.log(text);
+      } catch (e) {
+        console.log(e);
+      }
+    },
+    sendMessage() {
+      // To hide soft keyboard
+      Ember.$("textarea").trigger("blur");
+
+      this.set("inProgress", true);
+      var values = this.getProperties("body", "offer", "item", "isPrivate");
+      values.itemId = this.get("item.id");
+      values.offerId = this.get("offer.id");
+      values.createdAt = new Date();
+      values.sender = this.store.peekRecord(
+        "user",
+        this.get("session.currentUser.id")
+      );
+      this.get("messageLinkConvertor").convert(values);
+      var message = this.store.createRecord("message", values);
+      message
+        .save()
+        .then(() => {
+          this.set("body", "");
+        })
+        .catch(error => {
+          this.store.unloadRecord(message);
+          throw error;
+        })
+        .finally(() => this.set("inProgress", false));
```

Comment:
a non-returned promise with a `throw` means this is an uncaught and unhandled error